content/5_Plat/2DRQ.mdx

```diff
@@ -322,93 +344,89 @@ And you might use it like so:
 
 <LanguageSection>
 <CPPSection>
+
 ```cpp
-#include <algorithm>
-#include <cassert>
-#include <iostream>
-#include <vector>
+#include <bits/stdc++.h>
 using namespace std;
 using ll = long long;
 
-// BeginCodeSnip{BIT2D}
-// index of largest value <= x in v (sorted)
-// if v = [1, 2, 4], ind(v, 3) would return 1
-int ind(vector<int> v, int x) {
-	return upper_bound(v.begin(), v.end(), x) - v.begin() - 1;
-}
-
-class BIT2D {
+/**
+ * Offline 2D Fenwick Tree implementation.
+ * Note that n needs to be of a reasonable size, and
+ * all the updates/queries inputs are zero indexed.
+ */
+template <typename T> class OfflineBIT2D {
   private:
-	int n;  // max x-coordinate
-	vector<vector<int>> vals, bit;
+	const int n;
+	vector<vector<int>> vals;
+	vector<vector<T>> bit;
+
+	int ind(const vector<int> &v, int x) {
```

Comment:
what, returns the index of `x` in v?

Reply:
i should probably add a doc comment, but it returns the first index $i$ such that $v[i] <= x$

Reply:
yes please add doc comment

Reply:
already added it